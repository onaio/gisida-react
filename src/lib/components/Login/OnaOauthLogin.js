import React, { Component } from 'react';
import { useOAuthLogin } from '@onaio/gatekeeper/dist/components/login';
import { AuthorizationGrantType } from '@onaio/gatekeeper';

class OnaOauthLogin extends Component {
  constructor(props) {
    super(props);
    this.state = {
      authorizationUris: {},
    };
  }

  componentDidMount() {
    const clientID = process.env.REACT_APP_GISIDA_CANOPY_CLIENT_ID
    const redirectUri = location.protocol + '//' + location.host + '/callback';
    const providers = {
      onadata: {
        accessTokenUri: '',
        authorizationUri: 'https://api.ona.io/o/authorize/',
        clientId: clientID,
        redirectUri: redirectUri,
        scopes: ['read', 'write'],
        state: 'abc',
        userUri: 'https://api.ona.io/user',
      },
    };
    const options = {
      providers,
      authorizationGrantType: AuthorizationGrantType.IMPLICIT,
    };
    const authorizationUris = useOAuthLogin(options);
    this.setState({
      authorizationUris,
    });
  }

  getProviderKey() {
    return Object.keys(this.state.authorizationUris).filter(key => key === 'onadata')[0];
  }

  render() {
    const { publicPassword, publicUsername } = this.props;
    const providerKey = this.getProviderKey();

    return (
      <form className="login-form">
        <div className="form-group">
          {publicPassword && publicUsername ? (
            <div>
              <small>Username: {publicUsername}</small>
              <br />
              <small>Password: {publicPassword}</small>
            </div>
          ) : null}
        </div>
        <a
          className="btn btn-default center-block btn-block"
          href={this.state.authorizationUris[providerKey]}
        >
          Login
            </a>
      </form>
    );
  }
}

export default OnaOauthLogin
