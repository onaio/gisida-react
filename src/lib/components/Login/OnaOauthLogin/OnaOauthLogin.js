import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { useOAuthLogin, AuthorizationGrantType } from '@onaio/gatekeeper';
import { getURLSearchParams } from '../../../utils';

class OnaOauthLogin extends Component {
  constructor(props) {
    super(props);
    const { clientID } = this.props;
    let authorizationUris = {};
    // Include and query params in the callback URI if any e.g shared layers and style
    const searchParamsString = getURLSearchParams().toString();

    if (clientID) {
      const redirectUri = `${location.protocol}//${location.host}/callback${
        searchParamsString ? '?' + searchParamsString : ''
      }`;
      const providers = {
        onadata: {
          accessTokenUri: '',
          authorizationUri: 'https://api.ona.io/o/authorize/',
          clientId: clientID,
          redirectUri: redirectUri,
          scopes: ['read', 'write'],
          state: 'abc',
          userUri: 'https://api.ona.io/user',
        },
      };
      const options = {
        providers,
        authorizationGrantType: AuthorizationGrantType.IMPLICIT,
      };
      authorizationUris = useOAuthLogin(options);
    }

    this.state = {
      authorizationUris,
    };
  }

  render() {
    const { provider, publicPassword, publicUsername } = this.props;

    if (!this.props.clientID || !this.state.authorizationUris[provider]) {
      return null;
    }

    return (
      <form className="login-form">
        <div className="form-group">
          {publicPassword && publicUsername ? (
            <div>
              <small>Username: {publicUsername}</small>
              <br />
              <small>Password: {publicPassword}</small>
            </div>
          ) : null}
          <a
            className="btn btn-default center-block btn-block"
            href={this.state.authorizationUris[provider]}
          >
            Login
          </a>
        </div>
      </form>
    );
  }
}

OnaOauthLogin.defaultProps = {
  provider: 'onadata',
};

OnaOauthLogin.PropTypes = {
  clientID: PropTypes.string.isRequired,
  provider: PropTypes.string,
};

export default OnaOauthLogin;
