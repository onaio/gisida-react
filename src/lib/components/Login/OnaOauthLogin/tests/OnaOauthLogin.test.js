import { mount } from 'enzyme';
import React from 'react';
import toJson from 'enzyme-to-json';
import OnaOauthLogin from './../OnaOauthLogin';

describe('/src/lib/components/Login/OnaOauthLogin', () => {
    const props = {
        clientID: 'client_id',
        provider: 'onadata'
    }
    it('renders correctly', () => {
        const wrapper = mount(<OnaOauthLogin {...props} />)
        expect(toJson(wrapper)).toMatchSnapshot()
        wrapper.unmount()
    });

    it('renders null if client id is missing', () => {
        const wrapper = mount(<OnaOauthLogin />)
        expect(wrapper.isEmptyRender()).toBe(true)
        wrapper.unmount()
    });

    it('renders default oauth provider if provider is not provided', () => {
        const propsDefaultProvider = {
            clientID: 'client_id'
        }
        const wrapper = mount(<OnaOauthLogin  {...propsDefaultProvider} />)
        expect(toJson(wrapper)).toMatchSnapshot()
        wrapper.unmount()
    })
})