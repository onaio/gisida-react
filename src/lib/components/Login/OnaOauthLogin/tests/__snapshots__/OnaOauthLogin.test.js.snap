// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/src/lib/components/Login/OnaOauthLogin renders correctly 1`] = `
<OnaOauthLogin
  clientID="client_id"
  provider="onadata"
>
  <form
    className="login-form"
  >
    <div
      className="form-group"
    >
      <a
        className="btn btn-default center-block btn-block"
        href="https://api.ona.io/o/authorize/?client_id=client_id&redirect_uri=http%3A%2F%2Flocalhost%2Fcallback&scope=read%20write&response_type=token&state=abc"
      >
        Login
      </a>
    </div>
  </form>
</OnaOauthLogin>
`;

exports[`/src/lib/components/Login/OnaOauthLogin renders default oauth provider if provider is not provided 1`] = `
<OnaOauthLogin
  clientID="client_id"
  provider="onadata"
>
  <form
    className="login-form"
  >
    <div
      className="form-group"
    >
      <a
        className="btn btn-default center-block btn-block"
        href="https://api.ona.io/o/authorize/?client_id=client_id&redirect_uri=http%3A%2F%2Flocalhost%2Fcallback&scope=read%20write&response_type=token&state=abc"
      >
        Login
      </a>
    </div>
  </form>
</OnaOauthLogin>
`;
